### PR TITLE
Component object read fingerprints

### PR DESCRIPTION
**Pull request by weaverryan**
https://github.com/symfony/ux/pull/500

> 6) BACKEND: read childrenFingerprints data for GET (JSON-encoded query param) and POST (request body). Use this to compare to the NEW fingerprint value for each child. If not changed, render the child as an empty element (with only data-live-id). If changed, render an empty element, but with data-live-id, the new fingerprint data- value and the new data-live-props-value.

This is where it all comes together. As a reminder: 

A) Component parent renders & all children are given a "fingerprint" that is a checksum of the "input" passed to create that child
B) On parent re-render, an array containing [{ id: 'the-child-id', fingerprint: 'its fingerprint'}, ... ] is sent along with the parent component data.
C) We will (THIS PR) use that info to figure out if the "fingerprint" of any specific children has changed (i.e. we calculate the deterministic id of a child component, find its fingerprint using that id from the data above, and compare). If it has NOT, we effectively send back an empty element, which says to the frontend "leave the child alone". If the fingerprint DID change, we send back an empty element containing the new fingerprint and new props. The frontend child component will replace its props with the ones from the server and then issue an Ajax request to re-render itself.

Why not just re-render a child component from inside the parent Ajax request if the "fingerprint" for that child changed? Why wait for the child to make its own Ajax request? Because, when we return the new child props to the frontend, the child will "replace" its existing props with the new props but KEEP any existing data (i.e. writable props). So when it re-renders, it will now have the new props + any writable data, which may have changed since the was  originally rendered.

Cheers!

### CODE REVIEW

File: src/LiveComponent/src/EventListener/InterceptChildComponentRenderSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        $this->deterministicTwigIdCalculator->calculateDeterministicId(increment: true);
+
+        $newPropsFingerprint = $this->fingerprintCalculator->calculateFingerprint($event->getProps());
+
+        if ($childFingerprints[$deterministicId] === $newPropsFingerprint) {
+            // the props passed to create this child have *not* changed
+            // return an empty element so the frontend knows to keep the current child
+
+            $rendered = sprintf(
+                '<div data-live-id="%s"></div>',
```
Comment (kbond) [inline src/LiveComponent/src/EventListener/InterceptChildComponentRenderSubscriber.php:80]: Will this be a problem if the child component isn't wrapped in a `div`?

File: src/LiveComponent/src/Twig/DeterministicTwigIdCalculator.php [hunk 2]
```diff
@@ -56,8 +52,125 @@ public function calculateDeterministicId(): string
...
+    }
+
+    /**
+     * Adapted from Twig\Error\Error::guessTemplateInfo().
+     *
+     * Any differences are marked below.
+     *
+     * @return array{name: string, line: int}
+     */
+    private function guessTemplateInfo(): array
```
Comment (kbond) [inline src/LiveComponent/src/Twig/DeterministicTwigIdCalculator.php:75]: I didn't think it possibly to make this _more_ complex! 😆

**Review by kbond — APPROVED** [review]

File: src/LiveComponent/src/EventListener/InterceptChildComponentRenderSubscriber.php (hunk 1 above)
Comment (weaverryan) [reply src/LiveComponent/src/EventListener/InterceptChildComponentRenderSubscriber.php:80]: Excellent question! It should be ok. If the child element is actually a `span`, we actually account for this by correcting this `div` into a `span` right before we diff them (on the server, we have no idea what element to use, but once we're in JavaScript, we can find the child element in the current HTML and the "new" HTML via the `data-live-id`, then make them match).

The other potential weird thing (but JS tests confirm it's ok) is if the child element is, for example, a `<tr>`... where it's not legal to have a `<div>`. In other words, the new HTML would be `<table><div>...`. But, that would be wrong for just a moment (and only in memory, we haven't actually used the new HTML at this point) and then we would correct it to a `<tr>`.

So there *are* some gymnastics happening in the JS, but it seems to work.

File: src/LiveComponent/src/Twig/DeterministicTwigIdCalculator.php (hunk 2 above)
Comment (weaverryan) [reply src/LiveComponent/src/Twig/DeterministicTwigIdCalculator.php:75]: whoooooops :p 

I'm pretty excited to have Twig in Symfony... maybe this is yet another thing that could be "fixed"/ made available.

**Pull request merged**